File: benchmarks/select_vs_epoll_benchmark.py
#!/usr/bin/env python3
"""Benchmark select vs epoll and plot results with matplotlib.

This script creates many non-blocking pipes, triggers readability events, and
measures the time to observe and consume those events using two polling APIs.

Outputs:
- benchmarks/select_vs_epoll_results.csv
- benchmarks/select_vs_epoll_plot.png
"""

from __future__ import annotations

import argparse
import csv
import os
import random
import select
import statistics
import sys
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Callable

import matplotlib.pyplot as plt


@dataclass
class BenchmarkRow:
    backend: str
    fds: int
    iterations: int
    mean_us: float
    median_us: float
    p95_us: float


def percentile(sorted_values: list[float], p: float) -> float:
    if not sorted_values:
        return 0.0
    if p <= 0:
        return sorted_values[0]
    if p >= 100:
        return sorted_values[-1]
    k = (len(sorted_values) - 1) * (p / 100.0)
    lower = int(k)
    upper = min(lower + 1, len(sorted_values) - 1)
    frac = k - lower
    return sorted_values[lower] * (1.0 - frac) + sorted_values[upper] * frac


def _close_pipes(pairs: list[tuple[int, int]]) -> None:
    for rfd, wfd in pairs:
        for fd in (rfd, wfd):
            try:
                os.close(fd)
            except OSError:
                pass


def _create_pipe_pairs(n: int) -> tuple[list[tuple[int, int]], list[int], dict[int, int]]:
    pairs: list[tuple[int, int]] = []
    read_fds: list[int] = []
    writer_by_reader: dict[int, int] = {}

    for _ in range(n):
        rfd, wfd = os.pipe()
        os.set_blocking(rfd, False)
        os.set_blocking(wfd, False)
        pairs.append((rfd, wfd))
        read_fds.append(rfd)
        writer_by_reader[rfd] = wfd

    return pairs, read_fds, writer_by_reader


def run_select(nfds: int, iterations: int, timeout_s: float) -> list[float]:
    pairs, read_fds, writer_by_reader = _create_pipe_pairs(nfds)
    latencies_us: list[float] = []

    try:
        for _ in range(iterations):
            target = random.choice(read_fds)
            os.write(writer_by_reader[target], b"x")
            start_ns = time.perf_counter_ns()
            ready, _, _ = select.select(read_fds, [], [], timeout_s)
            elapsed_us = (time.perf_counter_ns() - start_ns) / 1000.0

            if not ready:
                raise RuntimeError("select() timed out waiting for readiness event")

            for rfd in ready:
                try:
                    os.read(rfd, 1)
                except BlockingIOError:
                    pass

            latencies_us.append(elapsed_us)

    finally:
        _close_pipes(pairs)

    return latencies_us


def run_epoll(nfds: int, iterations: int, timeout_s: float) -> list[float]:
    pairs, read_fds, writer_by_reader = _create_pipe_pairs(nfds)
    latencies_us: list[float] = []
    ep = select.epoll()

    try:
        for rfd in read_fds:
            ep.register(rfd, select.EPOLLIN)

        for _ in range(iterations):
            target = random.choice(read_fds)
            os.write(writer_by_reader[target], b"x")
            start_ns = time.perf_counter_ns()
            ready = ep.poll(timeout_s, maxevents=nfds)
            elapsed_us = (time.perf_counter_ns() - start_ns) / 1000.0

            if not ready:
                raise RuntimeError("epoll.poll() timed out waiting for readiness event")

            for rfd, _event_mask in ready:
                try:
                    os.read(rfd, 1)
                except BlockingIOError:
                    pass

            latencies_us.append(elapsed_us)

    finally:
        try:
            ep.close()
        except Exception:
            pass
        _close_pipes(pairs)

    return latencies_us


def summarize(backend: str, nfds: int, samples: list[float]) -> BenchmarkRow:
    sorted_samples = sorted(samples)
    return BenchmarkRow(
        backend=backend,
        fds=nfds,
        iterations=len(samples),
        mean_us=statistics.fmean(samples),
        median_us=statistics.median(sorted_samples),
        p95_us=percentile(sorted_samples, 95),
    )


def write_csv(rows: list[BenchmarkRow], path: Path) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("w", newline="", encoding="utf-8") as f:
        writer = csv.writer(f)
        writer.writerow(["backend", "fds", "iterations", "mean_us", "median_us", "p95_us"])
        for row in rows:
            writer.writerow(
                [
                    row.backend,
                    row.fds,
                    row.iterations,
                    f"{row.mean_us:.3f}",
                    f"{row.median_us:.3f}",
                    f"{row.p95_us:.3f}",
                ]
            )


def read_csv(path: Path) -> list[BenchmarkRow]:
    rows: list[BenchmarkRow] = []
    with path.open("r", newline="", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        for record in reader:
            rows.append(
                BenchmarkRow(
                    backend=record["backend"],
                    fds=int(record["fds"]),
                    iterations=int(record["iterations"]),
                    mean_us=float(record["mean_us"]),
                    median_us=float(record["median_us"]),
                    p95_us=float(record["p95_us"]),
                )
            )
    return rows


def plot(rows: list[BenchmarkRow], path: Path) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)

    by_backend: dict[str, list[BenchmarkRow]] = {}
    for row in rows:
        by_backend.setdefault(row.backend, []).append(row)

    plt.figure(figsize=(9, 5.5), dpi=150)
    for backend, backend_rows in sorted(by_backend.items()):
        backend_rows.sort(key=lambda r: r.fds)
        x = [r.fds for r in backend_rows]
        y = [r.median_us for r in backend_rows]
        plt.plot(x, y, marker="o", linewidth=2, label=f"{backend} median")

    plt.title("select vs epoll readiness latency")
    plt.xlabel("Number of monitored file descriptors")
    plt.ylabel("Latency per poll call (microseconds)")
    plt.grid(alpha=0.3)
    plt.legend()
    plt.tight_layout()
    plt.savefig(path)
    plt.close()


def default_fd_sizes() -> list[int]:
    return [32, 64, 128, 256, 512]


def run_backend(
    name: str,
    runner: Callable[[int, int, float], list[float]],
    fd_sizes: list[int],
    iterations: int,
    timeout_s: float,
) -> list[BenchmarkRow]:
    rows: list[BenchmarkRow] = []
    for nfds in fd_sizes:
        try:
            samples = runner(nfds, iterations, timeout_s)
            rows.append(summarize(name, nfds, samples))
        except (ValueError, OSError) as exc:
            print(f"Skipping {name} nfds={nfds}: {exc}", file=sys.stderr)
    return rows


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Benchmark select vs epoll and plot the results.")
    parser.add_argument(
        "--iterations",
        type=int,
        default=5000,
        help="Number of readiness events to sample per file descriptor size.",
    )
    parser.add_argument(
        "--timeout-ms",
        type=float,
        default=100.0,
        help="Poll timeout in milliseconds for each sample.",
    )
    parser.add_argument(
        "--fds",
        type=int,
        nargs="+",
        default=default_fd_sizes(),
        help="List of FD counts to benchmark (example: --fds 32 64 128 256).",
    )
    parser.add_argument(
        "--seed",
        type=int,
        default=42,
        help="Random seed for deterministic run ordering.",
    )
    parser.add_argument(
        "--csv-out",
        type=Path,
        default=Path("benchmarks/select_vs_epoll_results.csv"),
        help="Output CSV path.",
    )
    parser.add_argument(
        "--plot-out",
        type=Path,
        default=Path("benchmarks/select_vs_epoll_plot.png"),
        help="Output PNG path.",
    )
    parser.add_argument(
        "--plot-only-from-csv",
        type=Path,
        default=None,
        help="Skip Python benchmark execution and only plot rows from an existing CSV.",
    )
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    random.seed(args.seed)

    if args.plot_only_from_csv is not None:
        if not args.plot_only_from_csv.exists():
            print(f"CSV file does not exist: {args.plot_only_from_csv}", file=sys.stderr)
            return 4

        rows = read_csv(args.plot_only_from_csv)
        if not rows:
            print(f"CSV has no rows: {args.plot_only_from_csv}", file=sys.stderr)
            return 5

        plot(rows, args.plot_out)
        print(f"CSV:  {args.plot_only_from_csv}")
        print(f"Plot: {args.plot_out}")
        return 0

    if not hasattr(select, "epoll"):
        print("epoll is not available on this platform.", file=sys.stderr)
        return 1

    timeout_s = max(args.timeout_ms, 1.0) / 1000.0
    fd_sizes = sorted(set(x for x in args.fds if x > 0))
    if not fd_sizes:
        print("At least one positive FD count is required.", file=sys.stderr)
        return 2

    rows: list[BenchmarkRow] = []
    rows.extend(run_backend("select", run_select, fd_sizes, args.iterations, timeout_s))
    rows.extend(run_backend("epoll", run_epoll, fd_sizes, args.iterations, timeout_s))

    if not rows:
        print("No benchmark results were produced.", file=sys.stderr)
        return 3

    write_csv(rows, args.csv_out)
    plot(rows, args.plot_out)

    print("Benchmark summary (microseconds):")
    for row in sorted(rows, key=lambda r: (r.backend, r.fds)):
        print(
            f"{row.backend:>6} nfds={row.fds:>4} "
            f"mean={row.mean_us:8.3f} median={row.median_us:8.3f} p95={row.p95_us:8.3f}"
        )
    print(f"CSV:  {args.csv_out}")
    print(f"Plot: {args.plot_out}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: benchmarks/select_vs_epoll_benchmark.cc
#include <algorithm>
#include <array>
#include <chrono>
#include <cstdint>
#include <cstdlib>
#include <exception>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <numeric>
#include <optional>
#include <random>
#include <sstream>
#include <string>
#include <string_view>
#include <system_error>
#include <utility>
#include <vector>

#include <cerrno>

#include <fcntl.h>
#include <unistd.h>

#include "asyncio/backend/selector_epoll.h"
#include "asyncio/backend/selector_select.hh"

namespace {

struct Row {
  std::string backend;
  int fds{};
  int iterations{};
  double mean_us{};
  double median_us{};
  double p95_us{};
};

struct Options {
  int iterations{5000};
  double timeout_ms{100.0};
  std::vector<int> fds{32, 64, 128, 256, 512};
  std::string csv_out{"benchmarks/select_vs_epoll_results.csv"};
  uint32_t seed{42};
};

struct PipeSet {
  int read_fd{-1};
  int write_fd{-1};
};

void SetNonBlocking(int fd) {
  const int flags = ::fcntl(fd, F_GETFL, 0);
  if (flags < 0) {
    throw std::system_error(errno, std::system_category(), "fcntl(F_GETFL) failed");
  }

  if (::fcntl(fd, F_SETFL, flags | O_NONBLOCK) != 0) {
    throw std::system_error(errno, std::system_category(), "fcntl(F_SETFL) failed");
  }
}

std::vector<PipeSet> CreatePipes(int n) {
  std::vector<PipeSet> pipes;
  pipes.reserve(static_cast<size_t>(n));

  for (int i = 0; i < n; ++i) {
    std::array<int, 2> fds{-1, -1};
#if defined(O_CLOEXEC) && defined(O_NONBLOCK)
    if (::pipe2(fds.data(), O_CLOEXEC | O_NONBLOCK) != 0) {
      if (errno != ENOSYS) {
        throw std::system_error(errno, std::system_category(), "pipe2 failed");
      }
      if (::pipe(fds.data()) != 0) {
        throw std::system_error(errno, std::system_category(), "pipe failed");
      }
      SetNonBlocking(fds[0]);
      SetNonBlocking(fds[1]);
    }
#else
    if (::pipe(fds.data()) != 0) {
      throw std::system_error(errno, std::system_category(), "pipe failed");
    }
    SetNonBlocking(fds[0]);
    SetNonBlocking(fds[1]);
#endif
    pipes.push_back(PipeSet{fds[0], fds[1]});
  }

  return pipes;
}

void ClosePipes(const std::vector<PipeSet>& pipes) {
  for (const PipeSet& p : pipes) {
    if (p.read_fd >= 0) {
      ::close(p.read_fd);
    }
    if (p.write_fd >= 0) {
      ::close(p.write_fd);
    }
  }
}

double Percentile(const std::vector<double>& sorted_values, double p) {
  if (sorted_values.empty()) return 0.0;
  if (p <= 0.0) return sorted_values.front();
  if (p >= 100.0) return sorted_values.back();

  const double k = (static_cast<double>(sorted_values.size()) - 1.0) * (p / 100.0);
  const size_t lower = static_cast<size_t>(k);
  const size_t upper = std::min(lower + 1, sorted_values.size() - 1);
  const double frac = k - static_cast<double>(lower);
  return sorted_values[lower] * (1.0 - frac) + sorted_values[upper] * frac;
}

Row Summarize(std::string backend, int nfds, const std::vector<double>& samples) {
  std::vector<double> sorted = samples;
  std::sort(sorted.begin(), sorted.end());
  const double sum = std::accumulate(samples.begin(), samples.end(), 0.0);
  const double mean = samples.empty() ? 0.0 : sum / static_cast<double>(samples.size());
  const double median = sorted.empty() ? 0.0 : Percentile(sorted, 50.0);
  const double p95 = sorted.empty() ? 0.0 : Percentile(sorted, 95.0);

  return Row{
      .backend = std::move(backend),
      .fds = nfds,
      .iterations = static_cast<int>(samples.size()),
      .mean_us = mean,
      .median_us = median,
      .p95_us = p95,
  };
}

template <typename SelectorT>
std::vector<double> RunBackend(SelectorT& selector, int nfds, int iterations, double timeout_ms,
                               std::mt19937& rng) {
  std::vector<double> latencies_us;
  latencies_us.reserve(static_cast<size_t>(iterations));

  auto pipes = CreatePipes(nfds);
  std::vector<asyncio::IoEvent> events(static_cast<size_t>(nfds));

  for (const PipeSet& p : pipes) {
    selector.Register(p.read_fd, asyncio::IoEventFlags::kReadable, nullptr);
  }

  std::uniform_int_distribution<int> dist(0, nfds - 1);
  const auto timeout = std::chrono::duration_cast<std::chrono::nanoseconds>(
      std::chrono::duration<double, std::milli>(timeout_ms));

  try {
    for (int i = 0; i < iterations; ++i) {
      const int idx = dist(rng);
      const PipeSet& target = pipes[static_cast<size_t>(idx)];

      const ssize_t wrote = ::write(target.write_fd, "x", 1);
      if (wrote != 1) {
        throw std::system_error(errno, std::system_category(), "write to pipe failed");
      }

      const auto start = std::chrono::steady_clock::now();
      const int ready = selector.Select(events, timeout);
      const auto elapsed = std::chrono::steady_clock::now() - start;

      if (ready <= 0) {
        throw std::runtime_error("selector timed out waiting for readiness event");
      }

      for (int r = 0; r < ready; ++r) {
        const int fd = events[static_cast<size_t>(r)].handle;
        char byte = 0;
        for (;;) {
          const ssize_t n = ::read(fd, &byte, 1);
          if (n == 1) {
            break;
          }
          if (n < 0 && (errno == EAGAIN || errno == EWOULDBLOCK)) {
            break;
          }
          if (n < 0 && errno == EINTR) {
            continue;
          }
          if (n == 0) {
            break;
          }
          throw std::system_error(errno, std::system_category(), "read from pipe failed");
        }
      }

      latencies_us.push_back(
          static_cast<double>(std::chrono::duration_cast<std::chrono::nanoseconds>(elapsed).count()) /
          1000.0);
    }
  } catch (...) {
    for (const PipeSet& p : pipes) {
      try {
        selector.Unregister(p.read_fd);
      } catch (...) {
      }
    }
    ClosePipes(pipes);
    throw;
  }

  for (const PipeSet& p : pipes) {
    selector.Unregister(p.read_fd);
  }
  ClosePipes(pipes);
  return latencies_us;
}

void WriteCsv(const std::vector<Row>& rows, std::string_view path) {
  std::ofstream out{std::string(path)};
  if (!out.is_open()) {
    throw std::runtime_error("failed to open CSV output path");
  }

  out << "backend,fds,iterations,mean_us,median_us,p95_us\n";
  out << std::fixed << std::setprecision(3);
  for (const Row& row : rows) {
    out << row.backend << ',' << row.fds << ',' << row.iterations << ',' << row.mean_us << ','
        << row.median_us << ',' << row.p95_us << '\n';
  }
}

int ParsePositiveInt(const std::string& value, const char* flag_name) {
  try {
    const int parsed = std::stoi(value);
    if (parsed <= 0) {
      throw std::invalid_argument("non-positive");
    }
    return parsed;
  } catch (const std::exception&) {
    std::ostringstream oss;
    oss << "invalid " << flag_name << ": " << value;
    throw std::invalid_argument(oss.str());
  }
}

double ParsePositiveDouble(const std::string& value, const char* flag_name) {
  try {
    const double parsed = std::stod(value);
    if (parsed <= 0.0) {
      throw std::invalid_argument("non-positive");
    }
    return parsed;
  } catch (const std::exception&) {
    std::ostringstream oss;
    oss << "invalid " << flag_name << ": " << value;
    throw std::invalid_argument(oss.str());
  }
}

Options ParseArgs(int argc, char** argv) {
  Options opts;

  for (int i = 1; i < argc; ++i) {
    const std::string arg = argv[i];

    if (arg == "--iterations") {
      if (i + 1 >= argc) throw std::invalid_argument("--iterations requires a value");
      opts.iterations = ParsePositiveInt(argv[++i], "--iterations");
      continue;
    }

    if (arg == "--timeout-ms") {
      if (i + 1 >= argc) throw std::invalid_argument("--timeout-ms requires a value");
      opts.timeout_ms = ParsePositiveDouble(argv[++i], "--timeout-ms");
      continue;
    }

    if (arg == "--seed") {
      if (i + 1 >= argc) throw std::invalid_argument("--seed requires a value");
      opts.seed = static_cast<uint32_t>(ParsePositiveInt(argv[++i], "--seed"));
      continue;
    }

    if (arg == "--csv-out") {
      if (i + 1 >= argc) throw std::invalid_argument("--csv-out requires a value");
      opts.csv_out = argv[++i];
      continue;
    }

    if (arg == "--fds") {
      opts.fds.clear();
      while (i + 1 < argc && std::string_view(argv[i + 1]).rfind("--", 0) != 0) {
        opts.fds.push_back(ParsePositiveInt(argv[++i], "--fds value"));
      }
      if (opts.fds.empty()) {
        throw std::invalid_argument("--fds requires at least one value");
      }
      continue;
    }

    if (arg == "--help" || arg == "-h") {
      std::cout
          << "Usage: select_vs_epoll_benchmark_cpp [options]\n"
          << "  --iterations N        Samples per FD size (default 5000)\n"
          << "  --timeout-ms T        Poll timeout in milliseconds (default 100.0)\n"
          << "  --fds N1 N2 ...       FD sizes to test (default 32 64 128 256 512)\n"
          << "  --seed S              Random seed (default 42)\n"
          << "  --csv-out PATH        Output CSV path (default benchmarks/select_vs_epoll_results.csv)\n";
      std::exit(0);
    }

    throw std::invalid_argument("unknown argument: " + arg);
  }

  std::sort(opts.fds.begin(), opts.fds.end());
  opts.fds.erase(std::unique(opts.fds.begin(), opts.fds.end()), opts.fds.end());
  return opts;
}

int MainImpl(int argc, char** argv) {
  Options opts = ParseArgs(argc, argv);
  std::mt19937 rng(opts.seed);
  std::vector<Row> rows;

  for (int nfds : opts.fds) {
    try {
      asyncio::SelectSelector selector;
      rows.push_back(Summarize("select", nfds,
                               RunBackend(selector, nfds, opts.iterations, opts.timeout_ms, rng)));
    } catch (const std::exception& ex) {
      std::cerr << "Skipping select nfds=" << nfds << ": " << ex.what() << '\n';
    }
  }

#if defined(ASYNCIO_OS_LINUX)
  for (int nfds : opts.fds) {
    try {
      asyncio::EpollSelector selector;
      rows.push_back(Summarize("epoll", nfds,
                               RunBackend(selector, nfds, opts.iterations, opts.timeout_ms, rng)));
    } catch (const std::exception& ex) {
      std::cerr << "Skipping epoll nfds=" << nfds << ": " << ex.what() << '\n';
    }
  }
#else
  std::cerr << "epoll backend is unavailable on this platform\n";
#endif

  if (rows.empty()) {
    throw std::runtime_error("no benchmark rows produced");
  }

  WriteCsv(rows, opts.csv_out);

  std::sort(rows.begin(), rows.end(), [](const Row& a, const Row& b) {
    if (a.backend != b.backend) return a.backend < b.backend;
    return a.fds < b.fds;
  });

  std::cout << "Benchmark summary (microseconds):\n";
  std::cout << std::fixed << std::setprecision(3);
  for (const Row& row : rows) {
    std::cout << std::setw(6) << row.backend << " nfds=" << std::setw(4) << row.fds
              << " mean=" << std::setw(8) << row.mean_us << " median=" << std::setw(8)
              << row.median_us << " p95=" << std::setw(8) << row.p95_us << '\n';
  }
  std::cout << "CSV:  " << opts.csv_out << '\n';
  return 0;
}

}  // namespace

int main(int argc, char** argv) {
  try {
    return MainImpl(argc, argv);
  } catch (const std::exception& ex) {
    std::cerr << "Error: " << ex.what() << '\n';
    return 1;
  }
}
